include/inviwo/core/datastructures/geometry/typedmesh.h

```diff
@@ -0,0 +1,489 @@
+        return {static_cast<inviwo::BufferType>(attrib), location};
+    }
+
+#if _MSC_VER <= 1900
+    BufferTrait(type t) : t(t) {}
+    type t;
+    operator type() const { return t; }
+#endif
+
+    BufferTrait() {}
+    virtual ~BufferTrait() = default;
+
+    BufferTrait(Mesh &mesh) { mesh.addBuffer(bi(), buffer_); }
+
+    // Convenience functions
+    auto getTypedRAMRepresentation() const { return buffer_->getRAMRepresentation(); }
+    auto getTypedEditableRAMRepresentation() { return buffer_->getEditableRAMRepresentation(); }
+    auto &getTypedDataContainer() const { return getTypedRAMRepresentation()->getDataContainer(); }
+    auto &getTypedDataContainer() {
+        return getTypedEditableRAMRepresentation()->getDataContainer();
+    }
+
+    std::shared_ptr<Buffer<type>> buffer_{std::make_shared<Buffer<type>>()};
+};
+
+/**
+ * \ingroup typedmesh
+ * BufferTrait for Position buffers (glm::vec3)
+ */
+class PositionsBuffer : public BufferTrait<float, 3, static_cast<int>(BufferType::PositionAttrib)> {
+public:
+    using Base = BufferTrait<float, 3, static_cast<int>(BufferType::PositionAttrib)>;
+    using Base::Base;
+
+    std::shared_ptr<const Buffer<type>> getVertices() const { return Base::buffer_; }
+    std::shared_ptr<Buffer<type>> getEditableVertices() { return Base::buffer_; }
+
+    void setVertexPosition(size_t index, vec3 pos){
+        getEditableVertices()->getEditableRAMRepresentation()->set(index,pos);
+    }
+};
+
+/**
+ * \ingroup typedmesh
+ * BufferTrait for Normal buffers (glm::vec3)
+ */
+class NormalBuffer : public BufferTrait<float, 3, static_cast<int>(BufferType::NormalAttrib)> {
+public:
+    using Base = BufferTrait<float, 3, static_cast<int>(BufferType::NormalAttrib)>;
+    using Base::Base;
+
+    std::shared_ptr<const Buffer<type>> getNormals() const { return Base::buffer_; }
+    std::shared_ptr<Buffer<type>> getEditableNormals() { return Base::buffer_; }
+
+    void setVertexNormal(size_t index, vec3 normal){
+        getEditableNormals()->getEditableRAMRepresentation()->set(index,normal);
+    }
+};
+
+/**
+ * \ingroup typedmesh
+ * BufferTrait for Colors buffers (glm::vec4)
+ */
+class ColorsBuffer : public BufferTrait<float, 4, static_cast<int>(BufferType::ColorAttrib)> {
+public:
+    using Base = BufferTrait<float, 4, static_cast<int>(BufferType::ColorAttrib)>;
+    using Base::Base;
+
+    std::shared_ptr<const Buffer<type>> getColors() const { return Base::buffer_; }
+    std::shared_ptr<Buffer<type>> getEditableColors() { return Base::buffer_; }
+
+    void setVertexColor(size_t index, vec4 color){
+        getEditableColors()->getEditableRAMRepresentation()->set(index,color);
+    }
+};
+
+/**
+ * \ingroup typedmesh
+ * BufferTrait for Texture Coordinate buffers (glm::vec3)
+ */
+class TexcoordBuffer : public BufferTrait<float, 3, static_cast<int>(BufferType::TexcoordAttrib)> {
+public:
+    using Base = BufferTrait<float, 3, static_cast<int>(BufferType::TexcoordAttrib)>;
```

Comment:
would it be possible to make the number of components a template argument? I don't see a point in always using vec3 for tex coords.

Reply:
done locally, will be pushed next commit